src/CP/valueselection/learning/supervisedlearnedheuristic.jl

```diff
@@ -0,0 +1,143 @@
+    end
+
+    # Reset the agent, useful for things like recurrent networks
+    Flux.reset!(valueSelection.agent)
+
+    if valueSelection.trainMode
+        valueSelection.agent(RL.PRE_EPISODE_STAGE, env)
+    end
+end
+
+"""
+    (valueSelection::SupervisedLearnedHeuristic)(::DecisionPhase, model::CPModel, x::Union{Nothing, AbstractIntVar}, current_status::Union{Nothing, Symbol})
+
+Observe, store useful informations in the buffer with agent(POST_ACT_STAGE, ...) and take a decision with the call to agent(PRE_ACT_STAGE).
+The metrics that aren't updated in the StepPhase, which are more related to variables domains, are updated here. Once updated, they can
+be used in the other set_reward! function as the reward of the last action will only be collected in the RL.POST_ACT_STAGE.
+"""
+function (valueSelection::SupervisedLearnedHeuristic)(PHASE::Type{DecisionPhase}, model::CPModel, x::Union{Nothing,AbstractIntVar})
+    # domain change metrics, set before reward is updated
+    set_metrics!(PHASE, valueSelection.search_metrics, model, x)
+    set_reward!(PHASE, valueSelection, model)
+    model.statistics.lastVar = x
+    env = get_observation!(valueSelection, model, x)
+
+    if valueSelection.firstActionTaken
+        if valueSelection.trainMode
+            valueSelection.agent(RL.POST_ACT_STAGE, env) # get terminal and reward
+        end
+    else
+        valueSelection.firstActionTaken = true
+    end
+
+    # If a solution is available, we choose the value of the variable in the solution.
+    if valueSelection.trainMode && !isnothing(valueSelection.helpSolution) 
+        action = valueSelection.helpSolution[x.id]
+    else # Else we choose the action provided by the agent
+        action = valueSelection.agent(env) # Choose action
+    end
+    
+    if valueSelection.trainMode
+        # TODO: swap to async computation once in deployment
+        #@async valueSelection.agent(RL.PRE_ACT_STAGE, env, action) # Store state and action
+        valueSelection.agent(RL.PRE_ACT_STAGE, env, action)
+    end
+
+    return action_to_value(valueSelection, action, state(env), model)
+end
+
+"""
+    get_eta(valueSelection::SupervisedLearnedHeuristic)
+
+Get the current value of `eta` (η), which is the probability for the solver to calculate and provide a classic CP-generated solution to the agent.
+"""
+function get_eta(valueSelection::SupervisedLearnedHeuristic)
+    valueSelection.step += 1
+    if valueSelection.decay_steps == 0
+        return valueSelection.eta_init
+    end
+
+    step = valueSelection.step
+    if step <= valueSelection.warmup_steps
+        return valueSelection.eta_init
+    elseif step >= (valueSelection.warmup_steps + valueSelection.decay_steps)
+        return valueSelection.eta_stable
+    else
+        steps_left = valueSelection.warmup_steps + valueSelection.decay_steps - step
+        return valueSelection.eta_stable + steps_left / valueSelection.decay_steps * (valueSelection.eta_init - valueSelection.eta_stable)
```

Comment:
Do we want a linear decay like this or would an exponential one be more adapted?

Reply:
It would indeed be nice to be able to choose, but it doesn't seem to me to be absolutely essential at the moment.